pkg/storage/tsdb/tombstones.go

```diff
@@ -0,0 +1,321 @@
+	return &Tombstone{
+		RequestCreatedAt: requestTime,
+		StateCreatedAt:   stateTime,
+		StartTime:        startTime,
+		EndTime:          endTime,
+		Selectors:        selectors,
+		UserID:           userID,
+		RequestID:        requestId,
+		State:            state,
+	}
+}
+
+// Uploads a tombstone file to object sotre
+func WriteTombstoneFile(ctx context.Context, bkt objstore.Bucket, userID string, cfgProvider bucket.TenantConfigProvider, tombstone *Tombstone) error {
+	userBkt := bucket.NewUserBucketClient(userID, bkt, cfgProvider)
+
+	data, err := json.Marshal(tombstone)
+	if err != nil {
+		return errors.Wrap(err, "serialize tombstone")
+	}
+
+	// Add the state of the request using the filename extension
+	filename := tombstone.RequestID + ".json." + string(tombstone.State)
+
+	fullTombstonePath := path.Join(TombstonePath, filename)
+
+	// Check if the tombstone already exists. Could be the case the same request was made
+	// and is already in the middle of deleting series. Creating a new tombstone would restart
+	// the progress
+	tombstoneExists, err := TombstoneExists(ctx, userBkt, userID, tombstone.RequestID, tombstone.State)
+	if err != nil {
+		level.Warn(util_log.Logger).Log("msg", "unable to check if the same tombstone already exists", "user", userID, "requestID", tombstone.RequestID, "err", err)
+	} else if tombstoneExists {
+		return ErrTombstoneAlreadyExists
+	}
+
+	return errors.Wrap(userBkt.Upload(ctx, fullTombstonePath, bytes.NewReader(data)), "upload tombstone file")
+}
+
+func TombstoneExists(ctx context.Context, bkt objstore.BucketReader, userID string, requestID string, state BlockDeleteRequestState) (bool, error) {
+	// Add the state of the request using the filename extension
+	filename := requestID + ".json." + string(state)
+	fullTombstonePath := path.Join(TombstonePath, filename)
+
+	exists, err := bkt.Exists(ctx, fullTombstonePath)
+
+	if exists || err != nil {
+		return exists, err
+	}
+
+	return false, nil
+}
+
+func GetDeleteRequestsForUser(ctx context.Context, bkt objstore.Bucket, cfgProvider bucket.TenantConfigProvider, userID string, requestId string) (*Tombstone, error) {
+	userBucket := bucket.NewUserBucketClient(userID, bkt, cfgProvider)
+
+	var deleteRequest *Tombstone = nil
+
+	for _, state := range PossibleDeletionState {
+		filename := requestId + ".json." + string(state)
+
+		exists, err := TombstoneExists(ctx, userBucket, userID, requestId, state)
+		if err != nil {
+			return nil, err
+		}
+		if !exists {
+			continue
+		}
+
+		t, err := readTombstoneFile(ctx, userBucket, userID, path.Join(TombstonePath, filename))
+		if err != nil {
+			return nil, err
+		}
+
+		//checking if there exists more than one file for the same tombstone
+		// If it is the case, then should remove the one with the older state
+		if deleteRequest == nil {
```

Comment:
A simpler way to do this would be to use an array: `found`
Assuming PossibeDeletionState is sorted, the last element in the array would be the latest state

```
found := []Tombstone {}

for _, state := range PossibleDeletionState {
    filename := requestId + ".json." + string(state)

    exists, err := TombstoneExists(ctx, userBucket, userID, requestId, state)

    if exists {
        t, err := readTombstoneFile(ctx, userBucket, userID, path.Join(TombstonePath, filename))
        append(found, t)
    }
}

return found[len(found) - 1]

```

Reply:
Great suggestion, makes it a lot more clearer. Thanks!